keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,432 @@
+
+CSI Node Plugins that invoke members of the CSIProxyService API needs to check the versions supported by csi-proxy.exe. If a version mismatch is detected, it is recommended that the plugin author logs this clearly and exits from the process. For an operator, a version mismatch situation will manifest as the node daemonset for the plugin (whose version expectation is not satisfied) ending up with a zero pod count. Checking the log of an exited CSI Node Plugin container on a Windows node will indicate the version mismatch.
+
+Once compiled for Windows, container images based on Window Base images (like NanoServer) needs to be published and maintained. Container images targeting Linux nodes will need to be based on the desired Linux distro base image.
+
+New YAMLs for DaemonSets associated with CSI Node Plugins needs to be authored that will (1) target Windows nodes and (2) use Windows paths for domain socket related paths as illustrated in the User Story section above.
+
+### Risks and Mitigations
+
+Any pod on a Windows node can be configured to mount `\\.\pipe\csi-proxy` and perform privileged operations. Thus csi-proxy presents a potential security risk. To mitigate the risk, an admission webhook will need to be implemented and deployed that will reject all containers that bind mount `\\.\pipe\csi-proxy` as a hostPath volume but does not have `privileged` flag set in the pod's securityContext specification. This allows the `privileged` setting to be used for Windows pods as an indication the container will perform privileged operations. Other cluster-wide policies (e.g. PSP) that act on the `privileged` setting in a container's `securityContext` can enforce the same for CSI Node plugins targeting Windows nodes. Note that this does not in any way change how the `privileged` setting is used today for Linux nodes.
+
+## Design Details
+
+### Test Plan
+
+Unit tests will be added to verify Windows related enhancements in existing Kubernetes components mentioned above.
+
+All E2E storage tests covering CSI plugins will be ported to Windows workloads and successfully executed with above enhancements in place along with csi-proxy.exe.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- csi-proxy.exe supports v1beta1 version of the CSIProxyService API.
+- CSI Node Plugins for GCE PD, Azure Disk and Azure File can support Windows using v1beta1 version of the CSIProxyService API.
+- end-2-end tests in place involving CSI storage scenarios that exercise csi-proxy.exe and related enhancements.
+
+#### Beta -> GA Graduation
+
+- All in-tree storage plugins that implements support for Windows can use csi-proxy.exe along with other enhancements listed above to successfully deploy CSI plugins on Windows nodes.
+- csi-proxy.exe supports v1 stable version of the CSIProxyService API.
+- Successful usage of csi-proxy.exe with support for v1 version of CSIProxyService API in Windows nodes by at least two storage vendors.
```

Comment:
Not sure if you want to bundle graduation criteria with exact implementation details (maybe implementation might change?), instead it might be helpful to focus on what the end result would be and the ways we can validate those results? Not sure about this comment though, maybe we do want to speak about exact implementation details for more accuracy